text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+    // this code works for all mainstream platforms.
+    match_cfg! {
+        unix => foo_unix(),
+        windows => foo_windows(),
+    }
+}
+```
+
+As with many lints, the portability lint is *best effort*: it is not required to
+provide airtight guarantees about portability. In particular, the proposed
+definition in this RFC ignores trait implementations and dispatch. In practice,
+this hole shouldn't be too significant: it's pretty rare to implement a trait
+conditionally based on platform information, and the goal of this RFC is to
+*guide* code toward mainstream compatibility without imposing a heavy burden.
+
+With that overview in mind, let's dig into the details.
+
+## The lint definition
+
+The lint is structured somewhat akin to a type and effect system: roughly
+speaking, items that are labeled with a given `cfg` assumption can only be used
+within code making that same `cfg` assumption.
+
+More precisely, each item has a *portability*, consisting of all the
+lexically-nested uses of `cfg`, including those expressed via `match_cfg`, but
+excluding mentions of Cargo features. If there are multiple uses of `cfg`, the
+portability is taken to be their *conjunction*:
+
+```rust
+#[cfg(unix)]
+mod foo {
+    #[cfg(target_pointer_width = "32")]
+    fn bar() {
+        // the portability of `bar` is `all(unix, target_pointer_width = "32")`
+    }
+}
+```
+
+The lint is then straightforward to define at a high level: it walks over item
+definitions and checks that the item's portability is *narrower* than the
+portability of items it references or invokes. For example, `bar` in the above
+could invoke an item with portability `unix` and/or `target_pointer_width =
+"32"`, but not one with portability `linux`.
+
+To fully define the lint, though, we need to give more details about what
+"narrower" means, and how referenced item portability is determined.
+
+### Comparing portabilities
+
+**What does it mean for a portability to be narrower?** In general, portability
+is a logical expression, using the operators `all`, `any`, `not` on top of
+primitive expressions like `unix`. Portability `P` is narrower than portability
+`Q` if `P` *implies* `Q` as a logic formula.
+
+In general, comparing two portabilities is equivalent to solving SAT, an
+NP-complete problem -- a frightening prospect for a lint! However, in practice
+these comparisons are going to be relatively small and simple SAT instances,
+given the way that `cfg` is used. We can likely get away with a naive SAT
+implementation, perhaps with a handful of optimiziations specific to our
+use-case. In the limit, there are also many well-known techniques for solving
+SAT efficiently even on very large examples that arise in real-world usage.
+
+#### Known implications
+
+Another aspect of portability comparison is the relationship between things like
+`unix` and `linux`. In logical terms, we want to say that `linux` implies
+`unix`, for example.
```

Comment:
@Ericson2314 you can write a custom target which would set a user-specified target_os string (there can’t be multiple possible values though, so your point stands). While I do not think it would be a problem to support `key = value` cfgs, I would like to see expansion just for completeness sake.